### ISSUE TITLE
RESTful server spec

### ISSUE THREAD

**Issue by lijunsong** [lijunsong/pollen-rock#31]
Write the spec of the RESTful API so the front end has a spec to work on.

**Comment by jameskolce** [comment]
Are there any constraints regarding how the file tree is represented? Or am I free to come up with a spec from scratch?

**Comment by lijunsong** [comment]
Does the current one in pollen-rock work for you? All file system op is
under /fs. Spec is almost done. I can send out a diff this week.

**Comment by jameskolce** [comment]
Haven't tried yet, right now I just have been exploring the code. And I was thinking of doing something here, but if you have something I prefer to wait.

**Comment by lijunsong** [comment]
Are you talking about creating a server or querying the server? Your front
end can just query the server, right?

**Comment by jameskolce** [comment]
Perhaps we are talking about different things, let me explain what was in my mind: 

I have been exploring the Elm code to understand everything before starting implementing stuff, sometimes I just tend to review everything and see if I could come up with something better from scratch, that includes refactoring and changing the code organization. 

So while I was doing that, I was checking the mock API from the JSON server and was thinking about the structure of the data, just to see if there were things to improve in that structure before getting into the API communication in Elm. And because you were considering coming up with a new mock API I thought that I could do something there, but first I was wondering if there are any limitations with the data coming from the racket server since both have to be structurally the same.

Does that make sense?

Basically, I haven't touched anything regarding API communication in Elm yet. I just wanted to make sure the test data that we would use is optimal before doing something else.

----

By the way, have you thought about using GraphQL in Pollen Rock? I haven't used it, but sounds like an interesting approach to query data.

**Comment by lijunsong** [comment]
Make sense. So if you want to get this rolling quickly, just hook up the
server. You may end up working on the mock database for a long time without
getting anything meaningful out of it.

I am not going to spend time on mock API for the same reason.

—-
What problem can GraphGL solve for us?

**Comment by jameskolce** [comment]
Yeah, you are right. It could be better if I work with the real server instead, I could take that opportunity to learn some Racket too.

Regarding GraphQL, since you mentioned that it's complicated to represent a directoy in a RESTful way, I was wondering if perhaps GraphQL could help, but as I said, I haven't used it so I don't know, just wondering if you knew about it or something.

**Comment by lijunsong** [comment]
I think we’ve gotten a handle on that. Let’s get the project rolling (at
least first iteration) and worry about these later. This is a small
project, so changing things should be super easy.

The editor would be the first app using the new server API, and then Atom,
and then the emacs mode. I had my first version of the three editors, so I
am pretty sure the server would definitely work, and work a lot better this
time.

**Comment by jameskolce** [comment]
Alright! I'll try to get more into it these days. Also will let you know if I come up with some new things.

**Comment by lijunsong** [comment]
Pushed ae7d75cc13cf6adf0cd47616b062c286e3df5f29 to add draft version of the spec.

**Comment by lijunsong** [comment]
Merged
